cloudevents/work/clientbuilder.go

```diff
@@ -0,0 +1,157 @@
+	return &ClientHolderBuilder{
+		clientID:           clientID,
+		config:             config,
+		informerResyncTime: defaultInformerResyncTime,
+	}
+}
+
+// WithClusterName set the managed cluster name when building a  manifestwork client for an agent.
+func (b *ClientHolderBuilder) WithClusterName(clusterName string) *ClientHolderBuilder {
+	b.clusterName = clusterName
+	return b
+}
+
+// WithCodecs add codecs when building a manifestwork client based on cloudevents.
+func (b *ClientHolderBuilder) WithCodecs(codecs ...generic.Codec[*workv1.ManifestWork]) *ClientHolderBuilder {
+	b.codecs = codecs
+	return b
+}
+
+// WithInformerConfig set the ManifestWorkInformer configs. If the resync time is not set, the default time (10 minutes)
+// will be used when building the ManifestWorkInformer.
+func (b *ClientHolderBuilder) WithInformerConfig(
+	resyncTime time.Duration, options ...workinformers.SharedInformerOption) *ClientHolderBuilder {
+	b.informerResyncTime = resyncTime
+	b.informerOptions = options
+	return b
+}
+
+// NewClientHolder returns a ClientHolder for works.
+func (b *ClientHolderBuilder) NewClientHolder(ctx context.Context) (*ClientHolder, error) {
+	switch config := b.config.(type) {
+	case *rest.Config:
+		kubeWorkClientSet, err := workclientset.NewForConfig(config)
+		if err != nil {
+			return nil, err
+		}
+
+		factory := workinformers.NewSharedInformerFactoryWithOptions(kubeWorkClientSet, b.informerResyncTime, b.informerOptions...)
+
+		return &ClientHolder{
+			workClient:           kubeWorkClientSet.WorkV1(),
+			manifestWorkInformer: factory.Work().V1().ManifestWorks(),
+		}, nil
+	case *mqtt.MQTTOptions:
+		if len(b.clusterName) != 0 {
+			return b.newAgentClients(ctx, config)
+		}
+
+		//TODO build manifestwork clients for source
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("unsupported client configuration type %T", config)
+	}
+}
+
+func (b *ClientHolderBuilder) newAgentClients(ctx context.Context, config *mqtt.MQTTOptions) (*ClientHolder, error) {
+	workLister := &ManifestWorkLister{}
+	watcher := watcher.NewManifestWorkWatcher()
+	agentOptions := mqtt.NewAgentOptions(config, b.clusterName, b.clientID)
+	cloudEventsClient, err := generic.NewCloudEventAgentClient[*workv1.ManifestWork](
+		ctx,
+		agentOptions,
+		workLister,
+		ManifestWorkStatusHash,
+		b.codecs...,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	manifestWorkClient := agentclient.NewManifestWorkAgentClient(cloudEventsClient, watcher)
+	workClient := &internal.WorkV1ClientWrapper{ManifestWorkClient: manifestWorkClient}
+	workClientSet := &internal.WorkClientSetWrapper{WorkV1ClientWrapper: workClient}
+	factory := workinformers.NewSharedInformerFactoryWithOptions(workClientSet, b.informerResyncTime, b.informerOptions...)
+	informers := factory.Work().V1().ManifestWorks()
+	manifestWorkLister := informers.Lister()
+	namespacedLister := manifestWorkLister.ManifestWorks(b.clusterName)
```

Comment:
I am worried that we have multiple thread write cache.Store at the same time. Is cache.Store thread safe? What happens if a thread is updating the spec and another is updating the status or finallizer?

Reply:
the informer cache is a [thread safe store](https://github.com/kubernetes/client-go/blob/v0.26.1/tools/cache/store.go#L273), when we update the spec/status/finallizer, all of these actions will be handled by the [reflector](https://github.com/kubernetes/client-go/blob/v0.26.1/tools/cache/reflector.go#L530) , the reflector will call its store to [update](https://github.com/kubernetes/client-go/blob/v0.26.1/tools/cache/thread_safe_store.go#L221) them, so I think it should be thread safe call

Reply:
I think we should add some note here that client and informer is sharing the store in this implementation. We may need to revisit the implementation in the future. Ideally, the store should be only written from the server.

Reply:
OK, added a todo here